File: testcase2.js
// secret.js
const fs = require("fs");

// paste your JSON here
const data = {
  "keys": { "n": 10, "k": 7 },
  "1": { "base": "6", "value": "13444211440455345511" },
  "2": { "base": "15", "value": "aed7015a346d635" },
  "3": { "base": "15", "value": "6aeeb69631c227c" },
  "4": { "base": "16", "value": "e1b5e05623d881f" },
  "5": { "base": "8", "value": "316034514573652620673" },
  "6": { "base": "3", "value": "2122212201122002221120200210011020220200" },
  "7": { "base": "3", "value": "20120221122211000100210021102001201112121" },
  "8": { "base": "6", "value": "20220554335330240002224253" },
  "9": { "base": "12", "value": "45153788322a1255483" },
  "10": { "base": "7", "value": "1101613130313526312514143" }
};

function gcd(a,b){ a=a<0n?-a:a; b=b<0n?-b:b; while(b){ [a,b]=[b,a%b]; } return a; }
class Fr{
  constructor(n=0n,d=1n){ if(d<0n){n=-n;d=-d;} const g=gcd(n<0n?-n:n,d); this.n=n/g; this.d=d/g; }
  add(b){ return new Fr(this.n*b.d+b.n*this.d,this.d*b.d); }
  sub(b){ return new Fr(this.n*b.d-b.n*this.d,this.d*b.d); }
  mul(b){ return new Fr(this.n*b.n,this.d*b.d); }
  div(b){ return new Fr(this.n*b.d,this.d*b.n); }
}

function parseBig(baseStr, valStr){
  const b = BigInt(parseInt(baseStr,10));
  let r=0n;
  for(let ch of valStr){
    let v;
    if(ch>='0'&&ch<='9') v=BigInt(ch.charCodeAt(0)-48);
    else v=BigInt(ch.toLowerCase().charCodeAt(0)-87);
    r=r*b+v;
  }
  return r;
}

const k = data.keys.k;
let pts=[];
for(const key in data){
  if(key==="keys") continue;
  pts.push([BigInt(key), parseBig(data[key].base, data[key].value)]);
}
pts=pts.slice(0,k); // pick first k points

// Interpolation
let coeffs = Array(k).fill(0).map(_=>new Fr(0n,1n));
for(let i=0;i<k;i++){
  const [xi,yi]=pts[i];
  let numer=[new Fr(1n,1n)], denom=new Fr(1n,1n);
  for(let j=0;j<k;j++){
    if(i===j) continue;
    const [xj]=pts[j];
    const term=[new Fr(-xj,1n), new Fr(1n,1n)];
    let tmp=Array(numer.length+term.length-1).fill(0).map(_=>new Fr(0n,1n));
    for(let p=0;p<numer.length;p++) for(let q=0;q<term.length;q++)
      tmp[p+q]=tmp[p+q].add(numer[p].mul(term[q]));
    numer=tmp;
    denom=denom.mul(new Fr(xi-xj,1n));
  }
  const factor=new Fr(yi,1n).div(denom);
  for(let t=0;t<numer.length;t++) coeffs[t]=coeffs[t].add(numer[t].mul(factor));
}

const integers=coeffs.map(c=>{
  if(c.d!==1n) throw new Error("non-integer coeff");
  return c.n.toString();
}).reverse();

console.log("Polynomial coeffs (highest→lowest):", integers);
console.log("Secret (constant term):", integers[integers.length-1]);

File: testcase2.java
import java.math.BigInteger;
import java.util.*;

class Fraction {
    BigInteger n, d;

    public Fraction(BigInteger num, BigInteger den) {
        if (den.signum() < 0) {
            num = num.negate();
            den = den.negate();
        }
        BigInteger g = gcd(num.abs(), den.abs());
        n = num.divide(g);
        d = den.divide(g);
    }

    public Fraction add(Fraction b) {
        return new Fraction(n.multiply(b.d).add(b.n.multiply(d)), d.multiply(b.d));
    }

    public Fraction sub(Fraction b) {
        return new Fraction(n.multiply(b.d).subtract(b.n.multiply(d)), d.multiply(b.d));
    }

    public Fraction mul(Fraction b) {
        return new Fraction(n.multiply(b.n), d.multiply(b.d));
    }

    public Fraction div(Fraction b) {
        return new Fraction(n.multiply(b.d), d.multiply(b.n));
    }

    private static BigInteger gcd(BigInteger a, BigInteger b) {
        return b.equals(BigInteger.ZERO) ? a : gcd(b, a.mod(b));
    }
}

public class Secret {
    static BigInteger parseBig(String baseStr, String valStr) {
        BigInteger b = new BigInteger(baseStr);
        BigInteger r = BigInteger.ZERO;
        for (char ch : valStr.toCharArray()) {
            int v;
            if (ch >= '0' && ch <= '9') v = ch - '0';
            else v = Character.toLowerCase(ch) - 'a' + 10;
            r = r.multiply(b).add(BigInteger.valueOf(v));
        }
        return r;
    }

    public static void main(String[] args) {
        // JSON-like data
        Map<String, Map<String, String>> data = new HashMap<>();
        data.put("1", Map.of("base", "6", "value", "13444211440455345511"));
        data.put("2", Map.of("base", "15", "value", "aed7015a346d635"));
        data.put("3", Map.of("base", "15", "value", "6aeeb69631c227c"));
        data.put("4", Map.of("base", "16", "value", "e1b5e05623d881f"));
        data.put("5", Map.of("base", "8", "value", "316034514573652620673"));
        data.put("6", Map.of("base", "3", "value", "2122212201122002221120200210011020220200"));
        data.put("7", Map.of("base", "3", "value", "20120221122211000100210021102001201112121"));
        data.put("8", Map.of("base", "6", "value", "20220554335330240002224253"));
        data.put("9", Map.of("base", "12", "value", "45153788322a1255483"));
        data.put("10", Map.of("base", "7", "value", "1101613130313526312514143"));

        int k = 7; // from "keys.k"

        // Collect points
        List<BigInteger[]> pts = new ArrayList<>();
        for (String key : data.keySet()) {
            BigInteger x = new BigInteger(key);
            BigInteger y = parseBig(data.get(key).get("base"), data.get(key).get("value"));
            pts.add(new BigInteger[]{x, y});
        }

        // Sort by x to ensure consistent order, then take first k
        pts.sort(Comparator.comparing(a -> a[0]));
        pts = pts.subList(0, k);

        // Interpolation coefficients
        Fraction[] coeffs = new Fraction[k];
        for (int i = 0; i < k; i++) coeffs[i] = new Fraction(BigInteger.ZERO, BigInteger.ONE);

        for (int i = 0; i < k; i++) {
            BigInteger xi = pts.get(i)[0];
            BigInteger yi = pts.get(i)[1];

            List<Fraction> numer = new ArrayList<>();
            numer.add(new Fraction(BigInteger.ONE, BigInteger.ONE));
            Fraction denom = new Fraction(BigInteger.ONE, BigInteger.ONE);

            for (int j = 0; j < k; j++) {
                if (i == j) continue;
                BigInteger xj = pts.get(j)[0];

                // term = (x - xj)
                List<Fraction> term = Arrays.asList(
                        new Fraction(xj.negate(), BigInteger.ONE),
                        new Fraction(BigInteger.ONE, BigInteger.ONE)
                );

                // multiply numer * term
                List<Fraction> tmp = new ArrayList<>(Collections.nCopies(numer.size() + term.size() - 1,
                        new Fraction(BigInteger.ZERO, BigInteger.ONE)));

                for (int p = 0; p < numer.size(); p++) {
                    for (int q = 0; q < term.size(); q++) {
                        tmp.set(p + q, tmp.get(p + q).add(numer.get(p).mul(term.get(q))));
                    }
                }

                numer = tmp;
                denom = denom.mul(new Fraction(xi.subtract(xj), BigInteger.ONE));
            }

            Fraction factor = new Fraction(yi, BigInteger.ONE).div(denom);

            for (int t = 0; t < numer.size(); t++) {
                coeffs[t] = coeffs[t].add(numer.get(t).mul(factor));
            }
        }

        // Convert to integers
        List<BigInteger> integers = new ArrayList<>();
        for (Fraction c : coeffs) {
            if (!c.d.equals(BigInteger.ONE)) {
                throw new RuntimeException("Non-integer coefficient!");
            }
            integers.add(c.n);
        }
        Collections.reverse(integers);

        System.out.println("Polynomial coeffs (highest→lowest): " + integers);
        System.out.println("Secret (constant term): " + integers.get(integers.size() - 1));
    }
}
